cpp/src/Hypocenter.cpp

```diff
@@ -0,0 +1,226 @@
+	else
+		latitudeError = std::numeric_limits<double>::quiet_NaN();
+
+	// longitude error
+	if ((json.HasMember(LONGITUDE_ERROR_KEY) == true)
+			&& (json[LONGITUDE_ERROR_KEY].IsNumber() == true)
+			&& (json[LONGITUDE_ERROR_KEY].IsDouble() == true))
+		longitudeError = json[LONGITUDE_ERROR_KEY].GetDouble();
+	else
+		longitudeError = std::numeric_limits<double>::quiet_NaN();
+
+	// time error
+	if ((json.HasMember(TIME_ERROR_KEY) == true)
+			&& (json[TIME_ERROR_KEY].IsNumber() == true)
+			&& (json[TIME_ERROR_KEY].IsDouble() == true))
+		timeError = json[TIME_ERROR_KEY].GetDouble();
+	else
+		timeError = std::numeric_limits<double>::quiet_NaN();
+
+	// depth error
+	if ((json.HasMember(DEPTH_ERROR_KEY) == true)
+			&& (json[DEPTH_ERROR_KEY].IsNumber() == true)
+			&& (json[DEPTH_ERROR_KEY].IsDouble() == true))
+		depthError = json[DEPTH_ERROR_KEY].GetDouble();
+	else
+		depthError = std::numeric_limits<double>::quiet_NaN();
+}
+
+Hypocenter::Hypocenter(const Hypocenter & newHypocenter) {
+	latitude = newHypocenter.latitude;
+	longitude = newHypocenter.longitude;
+	depth = newHypocenter.depth;
+	time = newHypocenter.time;
+	latitudeError = newHypocenter.latitudeError;
+	longitudeError = newHypocenter.longitudeError;
+	depthError = newHypocenter.depthError;
+	timeError = newHypocenter.timeError;
+}
+
+Hypocenter::~Hypocenter() {
+}
+
+rapidjson::Value & Hypocenter::toJSON(
+		rapidjson::Value &json,
+		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
+	json.SetObject();
+
+	// required values
+	// latitude
+	if (std::isnan(latitude) != true)
+		json.AddMember(LATITUDE_KEY, latitude, allocator);
+
+	// longitude
+	if (std::isnan(longitude) != true)
+		json.AddMember(LONGITUDE_KEY, longitude, allocator);
+
+	// time
+	if (std::isnan(time) != true) {
+		std::string timestring = processingformats::ConvertEpochTimeToISO8601(
+				time);
+		rapidjson::Value timevalue;
+		timevalue.SetString(rapidjson::StringRef(timestring.c_str()),
+							allocator);
+		json.AddMember(TIME_KEY, timevalue, allocator);
+	}
+
+	// depth
+	if (std::isnan(depth) != true)
+		json.AddMember(DEPTH_KEY, depth, allocator);
+
+	// optional values
+	// latitude error
+	if (std::isnan(latitudeError) != true)
+		json.AddMember(LATITUDE_ERROR_KEY, latitudeError, allocator);
+
+	// longitude error
+	if (std::isnan(longitudeError) != true)
+		json.AddMember(LONGITUDE_ERROR_KEY, longitudeError, allocator);
+
+	// time error
+	if (std::isnan(timeError) != true)
+		json.AddMember(TIME_ERROR_KEY, timeError, allocator);
+
+	// depth error
+	if (std::isnan(depthError) != true)
+		json.AddMember(DEPTH_ERROR_KEY, depthError, allocator);
+
+	return (json);
+}
+
+std::vector<std::string> Hypocenter::getErrors() {
+	std::vector<std::string> errorlist;
+
+	// check required data
+	// latitude
+	if (std::isnan(latitude) == true) {
+		// latitude not found
+		errorlist.push_back("No Latitude in Hypocenter class.");
+	} else if ((latitude < -90) || (latitude > 90)) {
```

Comment:
be careful comparing doubles with integers, you might want to make all of these 90.00 etc. same long 180 and longitude and depth...